### PR TITLE
map_callrw implementation

### PR DESCRIPTION
**Pull request by Talkytitan5127**
add map_callrw implementation
Algorithm was taken from original vshard source

There is one moment that can be critical:
* that the var "refID" not global on level with another Go App using go-vshard-router [should be discussed]

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -200,7 +200,157 @@ func (r *Router) RouterCallImpl(ctx context.Context,
...
+		future := conn.Do(req, pool.RW)
+		futures[id] = future
+	}
+
+	// ref stage collect
+
+	totalBucketCount := uint16(0)
+
+	for _, future := range futures {
+		respData, err := future.Get()
```
Comment (ochaton) [inline api.go:263]: I don't like the synchronous way of getting futures. Maybe we will use list of channels of futures and will call reflect.Select on them?

File: vshard.go [hunk 2]
```diff
@@ -28,6 +28,7 @@ type Router struct {
 	}
 
 	knownBucketCount atomic.Int32
+	refID            atomic.Int64
```
Comment (KaymeKaydex) [inline vshard.go:31]: please write a small comment for this field

File: api.go [hunk 3]
```diff
@@ -200,7 +202,217 @@ func (r *Router) RouterCallImpl(ctx context.Context,
...
+		cases[i] = reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(metaData),
+		}
+	}
+
+	return cases
+}
+
+func (r *Router) callStorageUnref(refID int64) {
```
Comment (KaymeKaydex) [inline api.go:248]: please write a small comment for all functions

Comment (KaymeKaydex) [reply api.go:248]: u can use just a link to tarantool doc  with ref https://www.tarantool.io/ru/doc/latest/book/admin/vshard_admin/#bucket-ref

File: api.go [hunk 4]
```diff
@@ -2,7 +2,9 @@ package vshard_router
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"reflect"
```
Comment (KaymeKaydex) [inline api.go:7]: reflect usage sounds scary, cause it is very slow library, please describe why u used it, instead default libraries

File: api.go [hunk 5]
```diff
@@ -200,7 +202,217 @@ func (r *Router) RouterCallImpl(ctx context.Context,
...
+			}
+
+			return nil, err
+		}
+
+		id := metaData.replicaset.info.UUID
+		idToResult[id] = respData[1]
+	}
+
+	r.metrics().RequestDuration(time.Since(timeStart), true)
```
Comment (KaymeKaydex) [inline api.go:412]: i think we need one more label for this metric, like bool flag for map and non-map requests

File: vshard.go (hunk 2 above)
Comment (KaymeKaydex) [reply vshard.go:31]: u can use it from lua code https://github.com/tarantool/vshard/blob/master/vshard/router/init.lua#L60

File: api.go [hunk 6]
```diff
@@ -200,7 +202,217 @@ func (r *Router) RouterCallImpl(ctx context.Context,
...
+	Resp  []interface{}
+	Error error
+
+	replicaset *Replicaset
+	future     *tarantool.Future
+}
+
+func (r *Router) getMapCallSelectCases(req tarantool.Request) []reflect.SelectCase {
+	metaDatas := make([]<-chan *MapCallMetaData, 0)
+	for _, replicaset := range r.idToReplicaset {
```
Comment (KaymeKaydex) [inline api.go:215]: i think current logic looks  too hard and weight, in my mind -  the best way - to use just a slice with meta info; then u can send slice index  in range and ref to rs to goroutine with calll and return structs instead strange reflect struct

**Review by KaymeKaydex — CHANGES_REQUESTED** [review]
Thanks for the merge request, this is really important and currently unimplemented logic. It looks amazing, but I wrote a couple of notes about the logic, I think together we can make the code more optimal

File: api.go (hunk 5 above)
Comment (Talkytitan5127) [reply api.go:412]: make flag mapReduce

File: api.go (hunk 3 above)
Comment (Talkytitan5127) [reply api.go:206]: put comment

File: api.go [hunk 7]
```diff
@@ -200,7 +202,247 @@ func (r *Router) RouterCallImpl(ctx context.Context,
...
+		}
+
+		return nil
+	})
+
+	// map stage: collect
+
+	idToResult := make(map[uuid.UUID]interface{})
+
+	for i := 0; i < int(r.nWorkers); i++ {
```
Comment (KaymeKaydex) [inline api.go:376]: this worker pool model does not work

**Pull request merged**